store/migration/stream.go

```diff
@@ -0,0 +1,70 @@
+package migration
+
+import (
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	protoio "github.com/cosmos/gogoproto/io"
+	"github.com/cosmos/gogoproto/proto"
+
+	"cosmossdk.io/store/v2/snapshots"
+	snapshotstypes "cosmossdk.io/store/v2/snapshots/types"
+)
+
+// MigrationStream is a stream for migrating the whole state as a snapshot.
+// It's used to sync the whole state from the store/v1 to store/v2.
+// The main idea is to use the same snapshotter interface without writing to disk.
+type MigrationStream struct {
+	chBuffer chan proto.Message
+	err      atomic.Value // atomic error
+}
+
+var (
+	_ snapshots.WriteCloser = (*MigrationStream)(nil)
+	_ protoio.ReadCloser    = (*MigrationStream)(nil)
+)
+
+// NewMigrationStream returns a new MigrationStream.
+func NewMigrationStream(chBufferSize int) *MigrationStream {
+	return &MigrationStream{
+		chBuffer: make(chan proto.Message, chBufferSize),
+	}
+}
+
+// WriteMsg implements protoio.Write interface.
+func (ms *MigrationStream) WriteMsg(msg proto.Message) error {
+	ms.chBuffer <- msg
+	return nil
+}
+
+// CloseWithError implements snapshots.WriteCloser interface.
+func (ms *MigrationStream) CloseWithError(err error) {
+	ms.err.Store(err)
+	close(ms.chBuffer)
+}
+
+// ReadMsg implements protoio.Read interface.
+func (ms *MigrationStream) ReadMsg(msg proto.Message) error {
+	snapshotsItem, ok := msg.(*snapshotstypes.SnapshotItem)
+	if !ok {
+		return fmt.Errorf("unexpected message type: %T", msg)
+	}
+	item := <-ms.chBuffer
+	if item == nil {
+		return io.EOF
+	}
+	*snapshotsItem = *(item.(*snapshotstypes.SnapshotItem))
```

Comment:
I find this logic a bit confusing...we take in as input a `snapshotsItem` (casted from msg), but then we end up reading from a channel and assign it to the `snapshotsItem` variable? Can you allude to what/why we do this in the comments?

Reply:
ok, in snapshot flow, the writer and reader are separated since it assumes the state syncing between two nodes, in our case, we don't need to write/read the file, just need to pass the pointer, but the snapshotter interface requires the interface (proto.Message), especially in `Restore` function, it makes sense since we need to read the snapshot item from the file and deserialize. I just wanted to break the above flow and also re-use the same interface of snapshotter. That's why cast the interface to the concrete struct in this commit

Reply:
I would add sufficient go documentation alluding to why we do this then 👍

Reply:
There still isn't sufficient documentation here

Reply:
add more comments